### DISCUSSION TITLE
Nuxt 3, beforeRouteLeave in setup?

### DISCUSSION THREAD

**Discussion by jshimkoski** [nuxt/nuxt#17219] [Questions]
Is there a way to use the beforeRouteLeave lifecycle event inside of a setup script?

Or rather, is the `onBeforeRouteLeave` lifecycle composable missing? Seems this should exist.

**Comment by Tch1b0** [comment]
~I couldn't find a direct way to use the `onBeforeRouteLeave` event, so I think it really is missing.~
The event can be imported from the `vue-router` package, like shown in [the other answer](https://github.com/nuxt/framework/discussions/5830#discussioncomment-3148069).

## example workaround (not needed)

There does exist a `beforeResolve` function on the router, which I think has a similar effect.
So you could create a composable that contains the current callback for a component:

`composables/beforerouteleavehandler.js`
```js
export const useBeforeRouteLeaveHandler = () => {
  return useState("beforeRouteLeaveHandler", () => null)
}
```

And register the callback in your component:

`pages/<some-page>.vue`
```vue
<script setup>
useBeforeRouteLeaveHandler().value = () => {
  // code that should run before leaving the route
}
</script>
```

And then call this callback when a route is about to be resolved:

`app.vue`
```vue
<script setup>
useRouter().beforeResolve(() => {
  const routeHandler = useBeforeRouteLeaveHandler()
  if (routeHandler.value !== null) {
    // call the function contained in the routeHandler
    routeHandler.value()
    // reset the routeHandler
    routeHandler.value = null
  }
})
</script>
```

**But this isn't a replacement**, as in this example only one callback at a time can be applied, and the `beforeResolve` event does not work exactly like a `beforeRouteLeave` event.

**Comment by olivervorasai** [comment]
Here is an example of `onBeforeRouteLeave`

```vue
<script setup>
import { onBeforeRouteLeave } from 'vue-router';

onBeforeRouteLeave((to, from, next) => {
  alert('Leaving about page');
  next();
});
</script>

<template>
  <div>I Am About Page</div>
</template>
```

https://stackblitz.com/edit/github-f7fqpm?file=pages%2Fabout.vue

**Reply by jshimkoski** [reply]
I see that this works in the Stackblitz link, however, in my app (using the same code you provided), I see this warning:

```
Vue Router warn]: No active route record was found when calling `onBeforeRouteLeave()`. Make sure you call this function inside of a component child of <router-view>. Maybe you called it inside of App.vue?
```

I am calling it within the `/pages` directory in a page component so I'm not sure why it's losing its mind.

**Reply by PrimalDjinn** [reply]
i'm facing the same issue
